controllers/kubevirtmachine_controller.go

```diff
@@ -154,7 +157,11 @@ func (r *KubevirtMachineReconciler) Reconcile(goctx gocontext.Context, req ctrl.
 	}
 
 	// Handle non-deleted machines
-	return r.reconcileNormal(machineContext)
+	result, err := r.reconcileNormal(machineContext)
+	if err != nil || machineContext.KubevirtMachine.Spec.ProviderID == nil {
+		return result, err
+	}
+	return r.patchNode(machineContext)
```

Comment:
the checking of provider id here in this if def `if err != nil || machineContext.KubevirtMachine.Spec.ProviderID == nil` short circuits this function a bit. 

It's technically correct as the code is written today, but it's not obvious that the ProviderId is somehow related to the patchNode() function, and to not call patchNode if providerID == nil.

Can we move the providerID == nil check to the `r.patchNode` function so the logic can safely run like this.

```
	result, err := r.reconcileNormal(machineContext)
	if err != nil {
		return result, err
	}
	return r.patchNode(machineContext)

```


That said, I still feel weird about the way this logic is structured. There's a defer function that attempts to always patch the machine when the top level Reconcile function returns. Maybe we should be adding this patchNode there?

So, still with moving the check for providerID == nil into the patchNode function, something like this?

```
	// Always attempt to Patch the KubevirtMachine object and status after each reconciliation.
	defer func() {
		if err := machineContext.PatchKubevirtMachine(patchHelper); err != nil {
			machineContext.Logger.Error(err, "failed to patch KubevirtMachine")
			if rerr == nil {
				rerr = err
			}
		}

		if r.patchNode(machineContext)l err != nil {
			machineContext.Logger.Error(err, "failed to patch node")
			if rerr == nil {
				rerr = err
			}
		}
	}()
```

Reply:
Regarding the first comment:
I also added the `machineContext.KubevirtMachine.Spec.ProviderID == nil` validation, because there are cases in which the `reconcileNormal` function return in the middle without error. in this case need to return the result received from `reconcileNormal` fuction and return.
The way I have to validate this is by checking if `machineContext.KubevirtMachine.Spec.ProviderID` has valid value or not.

Regarding the second comment:
`PatchMachine` function is done in the `defer func` because this function (Reconcile) is responsible on reconciling the KubevirtMachine, therefore it no meter how it finished, need to update the KubevirtMachine with the changes.
This is not the same for `PatchNode`:
* This is done as workaround, because there is no Kubevirt cloud provider
* This update not need to happen always at the function end
* Must to remember that the relevant node might not be created yet, when this function is called
* If the Node doesn't exist, need to requeue the reconciler and try again

If the `PatchNode` name is confusing, it can be changed to `updateNode`, what do you say?

Reply:
> Regarding the first comment:
>I also added the machineContext.KubevirtMachine.Spec.ProviderID == nil validation, because there are cases in which the reconcileNormal function return in the middle without error. in this case need to return the result received from reconcileNormal fuction and return.
>The way I have to validate this is by checking if machineContext.KubevirtMachine.Spec.ProviderID has valid value or not.


yep, i get the reasoning. It's the naming of the functions and this very specific check for ProviderID that's a little strange. 

Looking at this code, how would I understand that `machineContext.KubevirtMachine.Spec.ProviderID` and `patchNode()` are related? I wouldn't draw that conclusion easily. 

If we have all the information we need in to patch the node with a providerID early, is there a reason why we can't call the original `SetProviderID()` function earlier in the reconcileNormal loop? It seems like we can call SetProviderID the moment that the VMI instance boots.

Reply:
I agree, it seems cleaner if we patched after each reconcile loop in defer function, per previous implementation, as opposed to invoking patch function explicitly, based on a specific condition.
patching logic should be generic and should handle all conditions.

Reply:
> If we have all the information we need in to patch the node with a providerID early, is there a reason why we can't call the original SetProviderID() function earlier in the reconcileNormal loop? It seems like we can call SetProviderID the moment that the VMI instance boots.

Its true that at the moment the VMI boot is done we have all the information to generate the ProviderID, but Node of this KubevirtMachine is not created yet
Therefore I changed the original `SetProviderID()` functions to only create the provider ID (and set it on the KubevirtMachine) and moved the Node update logic to another place, which will happen later, on another reconcile loop

Reply:
Hi @davidvossel @agradouski ,
Please see the following new commit https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/21/commits/08b66e93b6d893b4bd6db2af8e8cbef2f17f6771
* Renamed the function to `updateNodeProviderID`
* Added another status indicating if the node updated

Reply:
@nirarg this looks better to me
nit: maybe `r.updateNodeProviderID(machineContext)` should also be part of `r.reconcileNormal(machineContext)` routine?

Reply:
It doesn't feel right to me to have it in the reconcileNormal, as this action is not related to the KubevirtMachine reconciliation
Also, reconcileNormal is done only if the KubevirtMachine is not ready, which is not the condition to updateNodeProviderID, changing the location would result also with changing the condition
What do you say?

Reply:
@nirarg agree